### ISSUE TITLE
[torchdistx] Future of the large model initialization

### ISSUE THREAD

**Issue by YangFei1990** [pytorch/pytorch#96161]
For deep learning, when the model is large, model creation and initialization on host device will require tremendous time and sometimes causes host OOM. The existing [torchdistx](https://github.com/pytorch/torchdistx) package resolved this issue efficiently by introducing the `deferred_init` API which will first create model with fake tensor then materialized those tensor after the model is already partitioned, either by FSDP or other model parallelism methods.

However, it seems there is no active development in the torchdistx package, while the current FSDP implementation still depends on it for `deferred_init`. I'm wondering if there is any plan to merge the `deferred_init` API into the native Pytorch, and if no, what will be the native solution for large model initialization?

This issue is associated with this [PR](https://github.com/pytorch/xla/pull/4664) in the Pytorch/XLA repo.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @ezyang @eellison @bdhirsh @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu @saketh-are @soumith

**Comment by ngimel** [comment]
#91525 adds `torch.device` as a context manager, with that, you can initialize model on meta device

**Comment by YangFei1990** [comment]
> Note that there are two things here:
> 
> * If you want to initialize on the right device right away to speed up initialization / avoid memory issues. You can use `with torch.device("cuda:0")` for example to do that as suggested above.
> * If you want to skip init altogether because you're loading weights, `with torch.device("meta")` will skip the init and you can then load your weights in there (note that the loading is a bit tricky right now but you can ask if you need help with that).
> * If you want to delay the initialization, then you do need to use a more complex system as discussed by Ed above.
> 
> Do you have more details on which one you need exactly here?

torchdistx's `deferred_init` provides desired experience for large model initialization. Both methods provided above has drawbacks. The common experience for large model training is like this:
```
# Model is first create on host/CPU, usually from a third party library like HF
# Weight are already initialized
# This could be very slow or cause CPU OOM when the model is large
model = transformers.from_config(GPT3_CONFIG,...)

# Model is then wrapped with FSDP to shard the weights
model = FSDP(model, FSDP_CONFIG...)

# Move the sharded model to device
model.to(device)
```
The `meta` device approach does not record the operations on parameters, as a result, user needs to re-initialize the parameters again after materialize/moving to device. This introduces heavy burden for user and causes bug for FSDP( #90465 ). Torchdistx's `deferred_init`, on the other hand, can store all the operations for fake tensor, so user do not need to do anything.

**Comment by ezyang** [comment]
Yeah, this is the thing we should implement, just need to get someone to do it...

**Comment by learning-chip** [comment]
Is it possible to make a context manager like [deepspeed.zero.Init()](https://deepspeed.readthedocs.io/en/latest/zero3.html#constructing-massive-models) ?

**Comment by learning-chip** [comment]
Also in this blog post [Training a 1 Trillion Parameter Model With PyTorch Fully Sharded Data Parallel on AWS](https://medium.com/pytorch/training-a-1-trillion-parameter-model-with-pytorch-fully-sharded-data-parallel-on-aws-3ac13aa96cff):

> At the beginning of our experiment, 1T model initialization time took 3 hours. It initialized the full modules on CPUs and then passed CPU modules to the FSDP constructor for modules to be flattened and sharded on CPUs.
> 
> We reduced the initialization time from 3 hours to less than 1 minute by initializing the to-be-wrapped layers on GPUs directly and then passing these layers to the FSDP constructor to be flattened and sharded on GPUs. After this, the sharded layers are offloaded to CPUs.

However there is no mention for the source code of such implementation. Is it equivalent to deferred_init in Torchdistx?

**Comment by mikaylagawarecki** [comment]
Hey @learning-chip, would the following be helpful for you? On the nightly:

```
# don't load storages into RAM at load time
state_dict = torch.load('checkpoint.pth', mmap=True)
with torch.device('meta'):
     # don't allocate memory for tensor storages
    module = SomeModule()
# swap in parameters from checkpoint into module
module.load_state_dict(state_dict, assign=True)
```
